### PR TITLE
Transitory Sensitivity and Automatic Mechanism Analysis

### PR DESCRIPTION
**Pull request by mjohnson541**
This PR adds functionality for transitory sensitivities and also the tools for automatic mechanism analysis.

### CODE REVIEW

File: src/TransitorySensitivities.jl [hunk 1]
```diff
@@ -54,3 +54,64 @@ function splitnuminds(histo,n)
...
+    if length(fourinds) == 0
+        error("Could not find a multimodal slice")
+    end
+    ind = argmin((fracleftvals.- 0.5).^2 .+ (fracbelowvals .- 0.5).^2)
+    tau = histo.edges[1][fourinds[ind]+1]
+    return tau
+end
+
+function getdividingtimescale(sim,t;taumax=1e6,taumin=1e-18,taures=10.0^0.5,usediag=true)
+    Jy = jacobiany(sim.sol(t),sim.domain.p,t,sim.domain,[]);
```
Comment (hwpang) [inline src/TransitorySensitivities.jl:114]: I feel that we should provide some way to feed in the interfaces to extend this to open system in the future. Not necessary to do in this PR

File: src/TransitorySensitivities.jl [hunk 2]
```diff
@@ -465,3 +465,136 @@ function transitorysensitivitiesadjointexact(ssys::SystemSimulation,t,name;tau=N
...
+    else
+        prob = remake(ssys.sol.prob;u0=sim.sol(t),tspan=(0.0,tau));
+        soladj = solve(prob,solver,reltol=reltol,abstol=abstol);
+        ssysadj = SystemSimulation(soladj,[sim.domain for sim in ssys.sims],ssys.interfaces,ssys.p);
+
+        lm0 = zeros(length(sim.sol.prob.u0))
+        gu = ReverseDiff.gradient(x->soladj.prob.f(x,soladj.prob.p,0.0)[ind],soladj.prob.u0)
+        dlm(lm,p,t) = dlmdt(lm,p,t,simadj.sol.prob.f,gu,soladj)
+        ode = ODEProblem(dlm,lm0,(tau,0.0))
+        lmsol = solve(ode,solveradj,reltol=reltoladj,abstola=abstoladj);
```
Comment (hwpang) [inline src/TransitorySensitivities.jl:585]: Typo, should be `abstol`?

File: src/Domain.jl [hunk 3]
```diff
@@ -795,6 +795,28 @@ end
...
+    kfs = similar(y,length(d.phase.reactions))
+    krevs = similar(y,length(d.phase.reactions))
+    if !d.alternativepformat
+        kfs .= p[d.parameterindexes[1]-1+length(d.phase.species)+1:d.parameterindexes[1]-1+length(d.phase.species)+length(d.phase.reactions)]
+        Gs = p[d.parameterindexes[1]-1+1:d.parameterindexes[1]-1+length(d.phase.species)]
+    else
+        kfs .= d.p[length(d.phase.species)+1:end].*p[d.parameterindexes[1]-1+length(d.phase.species)+1:d.parameterindexes[1]-1+length(d.phase.species)+length(d.phase.reactions)]
+        Gs = d.p[1:length(d.phase.species)].+p[d.parameterindexes[1]-1+1:d.parameterindexes[1]-1+length(d.phase.species)]
+    end
+    krevs = getkfkrevs(d.phase,d.T,d.P,C,N,ns,Gs,d.diffusivity,V,0.0;kfs=kfs)[2]
```
Comment (hwpang) [inline src/Domain.jl:813]: Should we use `krevs .= ...` here?

**Review by hwpang — CHANGES_REQUESTED** [review]
@mjohnson541 Only several small request of changes. After addressing these, this PR can be merged!

File: src/TransitorySensitivities.jl (hunk 1 above)
Comment (mjohnson541) [reply src/TransitorySensitivities.jl:114]: Just left out the sim.interfaces XD

File: src/Domain.jl (hunk 3 above)
Comment (mjohnson541) [reply src/Domain.jl]: done

File: src/TransitorySensitivities.jl (hunk 2 above)
Comment (mjohnson541) [reply src/TransitorySensitivities.jl]: yep

**Comment by mjohnson541** [comment]
Okay I think all of those should be fixed. I also added FastGuassQuadrature to the dependencies and removed the manifest file (which probably shouldn't have been tracked by git in the first place).

**Comment by hwpang** [comment]
This PR looks good to me! I will go ahead and merge it.

**Pull request merged**